cedar-policy-core/src/authorizer/partial_response.rs

```diff
@@ -0,0 +1,481 @@
+            .satisfied_forbids
+            .iter()
+            .map(|(id, a)| (id, (&self.true_expr, a)));
+        let falses = self
+            .false_forbids
+            .iter()
+            .map(|(id, (_, a))| (id, (&self.false_expr, a)));
+        let nontrivial = self
+            .residual_forbids
+            .iter()
+            .map(|(id, (r, a))| (id, (r, a)));
+        trues.chain(falses).chain(nontrivial)
+    }
+
+    /// Return the residual for a given [`PolicyID`], if it exists in the response
+    pub fn get(&self, id: &PolicyID) -> Option<&Expr> {
+        self.get_true(id)
+            .or_else(|| self.get_false(id).or_else(|| self.get_residual(id)))
+    }
+
+    fn get_true(&self, id: &PolicyID) -> Option<&Expr> {
+        self.satisfied_permits
+            .get(id)
+            .or_else(|| self.satisfied_forbids.get(id))
+            .map(|_| &self.true_expr)
+    }
+
+    fn get_false(&self, id: &PolicyID) -> Option<&Expr> {
+        self.false_permits
+            .get(id)
+            .or_else(|| self.false_forbids.get(id))
+            .map(|_| &self.false_expr)
+    }
+
+    fn get_residual(&self, id: &PolicyID) -> Option<&Expr> {
+        self.residual_permits
+            .get(id)
+            .or_else(|| self.residual_forbids.get(id))
+            .map(|(r, _)| r)
+    }
+
+    /// Attempt to re-authorize this response given a mapping from unknowns to values
+    pub fn reauthorize(
+        &self,
+        mapping: &HashMap<SmolStr, Value>,
+        auth: &Authorizer,
+        r: Request,
+        es: &Entities,
+    ) -> Result<Self, PolicySetError> {
+        let policyset = self.all_policies(mapping)?;
+        Ok(auth.is_authorized_core(r, &policyset, es))
+    }
+
+    fn all_policies<'a>(
+        &'a self,
+        mapping: &'a HashMap<SmolStr, Value>,
+    ) -> Result<PolicySet, PolicySetError> {
+        PolicySet::try_from_iter(
+            build_policies(Effect::Permit, self.all_permit_residuals(), mapping).chain(
+                build_policies(Effect::Forbid, self.all_forbid_residuals(), mapping),
+            ),
+        )
+    }
+
+    fn errors(self) -> impl Iterator<Item = AuthorizationError> {
+        self.residual_forbids
+            .into_iter()
+            .chain(self.residual_permits)
+            .map(
+                |(id, (expr, _))| AuthorizationError::PolicyEvaluationError {
+                    id: id.clone(),
+                    error: EvaluationError::non_value(expr),
+                },
+            )
+            .chain(self.errors)
+    }
+}
+
+impl From<PartialResponse> for Response {
+    fn from(p: PartialResponse) -> Self {
+        let decision = if !p.satisfied_permits.is_empty() && p.satisfied_forbids.is_empty() {
+            Decision::Allow
+        } else {
+            Decision::Deny
+        };
+        Response::new(
+            decision,
+            p.definitely_determining().cloned().collect(),
+            p.errors().collect(),
+        )
+    }
+}
```

Comment:
for discussion: should this return `Option<Response>` and return `None` in the case where the partial decision is `None`?

Reply:
My vote: no
Reason: this goes against the error policies of Cedar, that errors cause a policy not to apply.

Reply:
I disagree with the interpretation of "nontrivial residuals are like errors".

In the Lean, the `PartialResponse.decision` method returns tri-state: Allow, Deny, or Unknown; and only returns Allow/Deny if that would be the decision for all possible substitutions of the unknowns.  This seems more intuitive to me.